cpp/src/arrow/array/builder_base.cc

```diff
@@ -92,6 +95,163 @@ Status ArrayBuilder::Advance(int64_t elements) {
   return null_bitmap_builder_.Advance(elements);
 }
 
+namespace {
+struct AppendScalarImpl {
+  template <typename T>
+  enable_if_t<has_c_type<T>::value || is_decimal_type<T>::value, Status> Visit(const T&) {
```

Comment:
This doesn't include `FixedSizeBinaryType`, which doesn't require a ReserveData call

Reply:
FixedSizeBinaryType didn't work with this specialization anyways, I'll go add the required Builder::Append(Buffer) call that's needed.